exp/etcdrestore/controllers/etcdmachinesnapshot_controller.go

```diff
@@ -60,18 +67,224 @@ func (r *EtcdMachineSnapshotReconciler) SetupWithManager(_ context.Context, mgr
+		Name:      req.Name,
+		Namespace: req.Namespace,
+	}}
+	if err := r.Client.Get(ctx, req.NamespacedName, etcdMachineSnapshot); apierrors.IsNotFound(err) {
+		// Object not found, return. Created objects are automatically garbage collected.
+		return ctrl.Result{}, nil
+	} else if err != nil {
+		log.Error(err, fmt.Sprintf("Unable to get etcdMachineSnapshot resource: %s", req.String()))
+		return ctrl.Result{}, err
+	}
+
+	// Handle deleted etcdMachineSnapshot
+	if !etcdMachineSnapshot.ObjectMeta.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, etcdMachineSnapshot)
+	}
+
+	// Ensure the finalizer is present
+	if !controllerutil.ContainsFinalizer(etcdMachineSnapshot, snapshotrestorev1.ETCDMachineSnapshotFinalizer) {
+		controllerutil.AddFinalizer(etcdMachineSnapshot, snapshotrestorev1.ETCDMachineSnapshotFinalizer)
+		if err := r.Client.Update(ctx, etcdMachineSnapshot); err != nil {
+			log.Error(err, "Failed to add finalizer to EtcdMachineSnapshot")
+			return ctrl.Result{}, err
+		}
+	}
+
+	return r.reconcileNormal(ctx, etcdMachineSnapshot)
 }
 
 func (r *EtcdMachineSnapshotReconciler) reconcileNormal(
-	_ context.Context, _ *snapshotrestorev1.EtcdMachineSnapshot,
-) (_ ctrl.Result, err error) {
+	ctx context.Context, etcdMachineSnapshot *snapshotrestorev1.EtcdMachineSnapshot,
+) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
+	// Handle different phases of the etcdmachinesnapshot creation process
+	switch etcdMachineSnapshot.Status.Phase {
+	case "":
+		// Initial phase, set to Pending
+		etcdMachineSnapshot.Status.Phase = snapshotrestorev1.ETCDSnapshotPhasePending
+		if err := r.Client.Status().Update(ctx, etcdMachineSnapshot); err != nil {
+			log.Error(err, "Failed to update EtcdMachineSnapshot status to Pending")
+			return ctrl.Result{}, err
+		}
+	case snapshotrestorev1.ETCDSnapshotPhasePending:
+		// Transition to Running
+		etcdMachineSnapshot.Status.Phase = snapshotrestorev1.ETCDSnapshotPhaseRunning
+		if err := r.Client.Status().Update(ctx, etcdMachineSnapshot); err != nil {
+			log.Error(err, "Failed to update EtcdMachineSnapshot status to Running")
+			return ctrl.Result{}, err
+		}
+	case snapshotrestorev1.ETCDSnapshotPhaseRunning:
+		// Check the status of the snapshot creation process
+		// List ETCDSnapshotFile resources to determine if the snapshot is complete
+		if err := checkSnapshotStatus(ctx, r, etcdMachineSnapshot); err != nil {
+			// If no matching ready snapshot is found, requeue the request
+			return ctrl.Result{RequeueAfter: 30 * time.Second}, nil
+		}
+		return ctrl.Result{}, nil
+	case snapshotrestorev1.ETCDSnapshotPhaseFailed:
+		// Handle snapshot creation failure
+		log.Error(nil, "Snapshot creation failed")
+		// Perform retry logic with exponential backoff
+		initialDelay := 1 * time.Second
+		maxRetries := 5
+
+		err := retryWithExponentialBackoff(ctx, initialDelay, maxRetries, func(ctx context.Context) error {
```

Comment:
maybe we can let the reconciler return an error instead? This way we'll use controller runtime to handle backoff.

Reply:
Yes, reconcile should never be blocking. It is always possible to just requeue. But I think returning error is not a good option, if this error is transient, as it might result in hitting very long wait periods, if the error persists for too long.

`RequeueAfter` IMO should be sufficient, unless this is the final error state, in which case returning a custom error LGTM too.

Reply:
I agree, it depends on how we want to handle it, in my opinion, we want to retry applying the system agent plan, and if it fails increase the timeout for reapplying to let users figure out possible issues.

Reply:
Updated to use `RequeueAfter`